src/azure-cli-core/azure/cli/core/extension/__init__.py

```diff
@@ -184,6 +184,24 @@ def get_azext_metadata(ext_dir):
                 azext_metadata = json.load(f)
         return azext_metadata
 
+    @staticmethod
+    def get_metadata_extras(ext_file):
+        metadata_extras = []
+        extra_const = "; extra == "
+        try:
+            if ext_file.endswith('.whl'):
+                ext_whl_metadata = pkginfo.Wheel(ext_file)
+            else:
+                raise ValueError()
+            # Parse out the extras for readability
+            for ext in ext_whl_metadata.requires_dist:
```

Comment:
This is the `METADATA` from https://github.com/Azure/azure-iot-cli-extension/releases/download/v0.14.0/azure_iot-0.14.0-py3-none-any.whl

```
Provides-Extra: uamqp
Requires-Dist: msrestazure (<2.0.0,>=0.6.3)
Requires-Dist: jsonschema (~=3.2.0)
Requires-Dist: azure-iot-device (~=2.5)
Requires-Dist: tqdm (~=4.62)
Requires-Dist: packaging
Requires-Dist: importlib-metadata; python_version < "3.8"
Provides-Extra: uamqp
Requires-Dist: uamqp (~=1.2); extra == 'uamqp'
```

1. Why does `Provides-Extra: uamqp` appear twice? Though it only appears once in https://github.com/Azure/azure-iot-cli-extension/blob/11637cd8a37f8194497ca28ba9518e7b04dc88c9/setup.py#L54
2. Per my understanding, this `get_metadata_extras` function parses extra names from `Requires-Dist`. However, the **extra name** denoted by `extra == ` may not be the same as the actual **package name**. For example, the extra name may be called `ua` while the actual package name is `uamqp`.
3. Why not use `Provides-Extra` to extract the extra names?

Reply:
From `pkginfo.Wheel` the metadata looks like this:
`
{
  "filename": "C:/Users/vilit/Downloads/azure_iot-0.14.0-py3-none-any.whl",
  "metadata_version": "2.0",
  "name": "azure-iot",
  "version": "0.14.0",
  "platforms": [
    "UNKNOWN"
  ],
  "summary": "The Azure IoT extension for Azure CLI.",
  "home_page": "https://github.com/azure/azure-iot-cli-extension",
  "author": "Microsoft",
  "author_email": "iotupx@microsoft.com",
  "license": "MIT",
  "classifiers": [
    "Development Status :: 4 - Beta",
    "Intended Audience :: Developers",
    "Intended Audience :: System Administrators",
    "Programming Language :: Python",
    "Programming Language :: Python :: 3",
    "Programming Language :: Python :: 3.6",
    "Programming Language :: Python :: 3.7",
    "Programming Language :: Python :: 3.8",
    "Programming Language :: Python :: 3.9",
    "Programming Language :: Python :: 3.10",
    "License :: OSI Approved :: MIT License"
  ],
  "requires_python": ">=3.6,<4",
  "requires_dist": [
    "msrestazure (<2.0.0,>=0.6.3)",
    "jsonschema (~=3.2.0)",
    "azure-iot-device (~=2.5)",
    "tqdm (~=4.62)",
    "packaging",
    "importlib-metadata; python_version < \"3.8\"",
    "uamqp (~=1.2); extra == 'uamqp'"
  ],
  "description": "The Azure IoT extension for Azure CLI. Intended for power users and/or automation of IoT solutions at scale.\n\n\n"
}
`

1. Not sure why - seems like a weird bug on the metadata side. Either way, this metadata file is generated _after_ installation. Once you hit this point in the code, the metadata file does not exist yet.
2. In `requires_dist` the extra name is what we want to use, not the actual package name (especially since we can map one extra name to many packages). So, in this case, if the extra name is bar as in `uamqp (~=1.2); extra == 'bar'`, we would want to extract 'bar' so we can install the package with azure_iot-0.14.0-py3-none-any.whl[bar]
3. See 1 - I am looking at the metadata before the package is installed using `pkginfo.Wheel`. This only provides `requires_dist` so I am using that.

Reply:
> Either way, this metadata file is generated after installation.

This is not true, this `METADATA` file is in the wheel https://github.com/Azure/azure-iot-cli-extension/releases/download/v0.14.0/azure_iot-0.14.0-py3-none-any.whl before installation.

`provides_extras` is empty for `azure_iot-0.14.0-py3-none-any.whl`.

![image](https://user-images.githubusercontent.com/4003950/164133026-3e0366f4-721f-45ac-8150-6adb9f0bbae1.png)

Meanwhile, `provide_extras` is correct for https://files.pythonhosted.org/packages/2d/61/08076519c80041bc0ffa1a8af0cbd3bf3e2b62af10435d269a9d0f40564d/requests-2.27.1-py2.py3-none-any.whl

![image](https://user-images.githubusercontent.com/4003950/164132967-fb0386e9-e55f-4804-a0d7-d54d6cb72864.png)

Looks like it is because `azure_iot` wheel has 

```
Metadata-Version: 2.0
```

but `requests` wheel has

```
Metadata-Version: 2.1
```

See

- https://peps.python.org/pep-0566/
- https://github.com/pypa/wheel/issues/231

Reply:
To add onto the above - the wheel package, which determines the metadata version, needs to be at least 0.31.0 to use metadata version 2.1.

azdev currently enforces wheel == 0.30.0 https://github.com/Azure/azure-cli-dev-tools/blob/4f97af4466e70496b810c811cc094729dabe2a0a/setup.py#L79

This means that most azure extensions will use the older metadata version of 2.0